lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,243 @@
+            recurse_data = from_data.kind_of?(Hash) ? from_data.values : from_data
+            recurse_data.each do |value|
+              get_file_contents(value, base_url)
+            end
+          end
+
+          # I'm on a Hash, process it.
+          if from_data.kind_of?(Hash)
+            from_data.each do |key, value|
+              next if ignore_if(key, value)
+              Fog::Logger.debug("Inspecting #{key}, #{value} at #{base_url}")
+
+              # Resolve relative paths.
+              base_url += '/' if base_url && !base_url.end_with?('/')
+              str_url = url_join(base_url, value)
+
+              next if @files.key?(str_url)
+              # Don't process file:// outside our base_url. TODO raise an exception here?
+              next if file_outside_base_url?(base_url, str_url)
+
+              file_content = get_content(str_url)
+
+              # get_file should not recurse hot templates.
+              if key == "type" && template_is_raw?(file_content) && !(@visited[str_url])
+                template = get_template_contents(str_url)
+                file_content = YAML.dump(template)
+              end
+
+              @files[str_url] = file_content
+              # replace the data value with the normalised absolute URL
+              Fog::Logger.debug("Replacing #{key} with #{str_url} in #{from_data}")
+              from_data[key] = str_url
+            end
+          end
+        end
+
+        private
+
+        # Retrive the content of a local or remote file.
+        #
+        # @param A local or remote uri.
+        #
+        # @raise ArgumentError if it's not a valid uri
+        #
+        # Protect open-uri from malign arguments like
+        #  - "|ls"
+        #  - multiline strings
+        def get_content(uri_or_filename)
+          remote_schemes = %w[http https ftp]
+          Fog::Logger.debug("Opening #{uri_or_filename}")
+
+          begin
+            # Validate URI to protect from open-uri attacks.
+            url = URI(uri_or_filename)
+
+            # Remote schemes must contain an host.
+            raise ArgumentError if url.host.nil? && remote_schemes.include?(url.scheme)
+
+            # Encode URI with spaces.
+            uri_or_filename = URI.encode(URI.decode(URI(uri_or_filename).to_s))
+          rescue URI::InvalidURIError
+            raise ArgumentError, "Not a valid URI: #{uri_or_filename}"
+          end
+
+          # TODO: A future revision may implement a retry.
+          # TODO: A future revision may limit download size.
+          content = ''
+          # open-uri doesn't open "file:///" uris.
+          uri_or_filename = uri_or_filename.sub(/^file:/, "")
+
+          open(uri_or_filename) { |f| content = f.read }
+          content
+        end
+
+        # Return true if the file is an heat template, false otherwise.
+        def template_is_raw?(content)
+          htv = content.index("heat_template_version:")
+          htv && htv < 5
```

Comment:
better to strip off leading white spaces and test whether content starts with "heat_template_version"